File: cplusplus/src/base/models/rectangle.cpp
#include <shapes/base/models.hpp>

namespace shapes::base::models {
Rectangle::Rectangle(double width, double height) {
    const double dimOne = (width > 0) ? width : 1;
    const double dimTwo = (height > 0) ? height : 1;
    this->width = dimOne;
    this->height = dimTwo;
}

auto Rectangle::area() const -> double {
    return width * height;
}

auto Rectangle::perimeter() const -> double {
    return 2 * (width + height);
}
} // namespace shapes::base::models

File: go/base/models/rectangle.go
package models

type Rectangle struct {
	width  float64
	height float64
}

func (Rectangle) New(width, height float64) *Rectangle {
	var dimOne = map[bool]float64{true: width, false: 1}[width > 0]
	var dimTwo = map[bool]float64{true: height, false: 1}[height > 0]
	return &Rectangle{width: dimOne, height: dimTwo}
}

func (r Rectangle) Area() float64 {
	return r.width * r.height
}

func (r Rectangle) Perimeter() float64 {
	return 2 * (r.width + r.height)
}
